docs/azure/sdk/unit-testing-mocking.md

```diff
@@ -0,0 +1,420 @@
+---
+title: Unit testing and mocking with the Azure SDK .NET
+description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
+ms.custom: devx-track-dotnet, engagement-fy23
+ms.date: 07/05/2023
+---
+
+# Unit testing and mocking with the Azure SDK for .NET
+
+Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls to Azure resources. Tests that run against live services can experience issues, such as, latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.
+
+In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/), which is a popular mocking and testing library for .NET.
+
+## Understand service clients
+
+A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it's important to be able to create an instance of the client that behaves as expected without making any network calls.
+
+Each of the Azure SDK clients follows mocking guidelines that allow their behavior to be overridden:
+
+* Each client offers at least one protected constructor to allow inheritance for testing.
+* All public client members are virtual to allow overriding.
+
+> [!NOTE]
+> The code examples in this article use types from the [`Azure.Security.KeyVault.Secrets`](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.
+
+To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).
```

Comment:
💡 I would consider both options here code smells. Where possible tests should be operating under production code to the maximum degree possible. When this involves hosted APIs that are not available (or not practical) during test scenarios, the preferred approach is to spin up a stub implementation of the actual service using something like self-hosted OWIN, and direct the production code to use that local endpoint as the target for its operations.

Reply:
In the complete unit test example at the end of the article the only part of the code that is being mocked or substituted are the SDK classes - the production code written for the app by the developer is still being tested. The SDK dependencies are being swapped with mocks because a unit test wouldn't be concerned with testing the code in external library dependencies - this was the guidance provided by the Azure SDK team when this content was requested.

Reply:
➡️ Will discuss with content authors internally to find a documentation plan that naturally encourages the use of tools for specific areas where they excel, and can't be interpreted as suggesting and single tool as a one-size-fixes-all solution.

Reply:
Hi @sharwell - I'd like to better understand your arguments.

This article isn't trying to sell Moq as a "one-size-fits-all solution", it's celebrating it as an OSS project that works well with the Azure .NET SDK. It's got nearly 500m downloads on NuGet and is used by 175k other public GitHub repositories for a reason.

Moq is used in the dotnet/runtime too, and it's perfectly valid when you know how to use the tool. The purpose of this article is to demonstrate how to use it.

What specifically about this article is off putting, or what specifically would you suggest changing? It's based on this blog post: https://devblogs.microsoft.com/azure-sdk/unit-testing-and-mocking.

Reply:
@sharwell this article follows our Azure SDK for .NET guidelines: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-mocking /cc @KrzysztofCwalina